File: packages/sdk-go/pkg/parallax/grpc_agent.go
package parallax

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"

	confidence "parallax/sdk-go/generated"
	registry "parallax/sdk-go/generated"
)

// Agent defines the interface that all Parallax agents must implement
type Agent interface {
	// GetID returns the agent's unique identifier
	GetID() string

	// GetName returns the agent's human-readable name
	GetName() string

	// GetCapabilities returns the agent's capabilities
	GetCapabilities() []string

	// GetMetadata returns optional agent metadata
	GetMetadata() map[string]string

	// Analyze performs the agent's main analysis task
	Analyze(ctx context.Context, task string, data interface{}) (*AnalyzeResult, error)

	// CheckHealth returns the agent's health status
	CheckHealth(ctx context.Context) (*HealthStatus, error)
}

// AnalyzeResult represents the result of an agent's analysis
type AnalyzeResult struct {
	Value         interface{}
	Confidence    float64
	Reasoning     string
	Uncertainties []string
	Metadata      map[string]string
}

// HealthStatus represents an agent's health
type HealthStatus struct {
	Status  string // "healthy", "degraded", "unhealthy"
	Message string
}

// GrpcAgent wraps an Agent with gRPC server capabilities
type GrpcAgent struct {
	confidence.UnimplementedConfidenceAgentServer
	
	agent        Agent
	server       *grpc.Server
	port         int
	registryAddr string
	leaseID      string
	stopCh       chan struct{}
	wg           sync.WaitGroup
	mu           sync.Mutex
}

// NewGrpcAgent creates a new gRPC-enabled agent
func NewGrpcAgent(agent Agent) *GrpcAgent {
	registryAddr := os.Getenv("PARALLAX_REGISTRY")
	if registryAddr == "" {
		registryAddr = "localhost:50051"
	}

	return &GrpcAgent{
		agent:        agent,
		registryAddr: registryAddr,
		stopCh:       make(chan struct{}),
	}
}

// Serve starts the gRPC server and registers with the control plane
func (g *GrpcAgent) Serve(port int) error {
	// Create gRPC server
	g.server = grpc.NewServer()
	confidence.RegisterConfidenceAgentServer(g.server, g)

	// Listen on port
	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		return fmt.Errorf("failed to listen: %w", err)
	}

	// Get actual port
	g.port = lis.Addr().(*net.TCPAddr).Port
	log.Printf("Agent %s (%s) listening on port %d", g.agent.GetName(), g.agent.GetID(), g.port)

	// Start server in goroutine
	g.wg.Add(1)
	go func() {
		defer g.wg.Done()
		if err := g.server.Serve(lis); err != nil {
			log.Printf("Failed to serve: %v", err)
		}
	}()

	// Register with control plane
	if err := g.register(); err != nil {
		log.Printf("Failed to register with control plane: %v", err)
		// Continue running even if registration fails
	}

	return nil
}

// Stop gracefully shuts down the agent
func (g *GrpcAgent) Stop() error {
	close(g.stopCh)

	// Unregister from control plane
	if err := g.unregister(); err != nil {
		log.Printf("Failed to unregister: %v", err)
	}

	// Stop gRPC server
	g.server.GracefulStop()
	
	// Wait for goroutines
	g.wg.Wait()

	return nil
}

// register registers the agent with the control plane
func (g *GrpcAgent) register() error {
	conn, err := grpc.NewClient(g.registryAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return fmt.Errorf("failed to connect to registry: %w", err)
	}
	defer conn.Close()

	client := registry.NewRegistryClient(conn)
	
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	req := &registry.RegisterRequest{
		Agent: &registry.AgentRegistration{
			Id:           g.agent.GetID(),
			Name:         g.agent.GetName(),
			Endpoint:     fmt.Sprintf("localhost:%d", g.port),
			Capabilities: g.agent.GetCapabilities(),
		},
	}

	resp, err := client.Register(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to register: %w", err)
	}

	g.mu.Lock()
	g.leaseID = resp.LeaseId
	g.mu.Unlock()

	log.Printf("Agent %s registered with control plane, lease_id: %s", g.agent.GetID(), g.leaseID)

	// Start lease renewal
	g.wg.Add(1)
	go g.renewLeaseLoop()

	return nil
}

// renewLeaseLoop periodically renews the agent's lease
func (g *GrpcAgent) renewLeaseLoop() {
	defer g.wg.Done()

	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	conn, err := grpc.NewClient(g.registryAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		log.Printf("Failed to connect for lease renewal: %v", err)
		return
	}
	defer conn.Close()

	client := registry.NewRegistryClient(conn)

	for {
		select {
		case <-g.stopCh:
			return
		case <-ticker.C:
			g.mu.Lock()
			leaseID := g.leaseID
			g.mu.Unlock()

			if leaseID == "" {
				continue
			}

			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			req := &registry.RenewRequest{LeaseId: leaseID}
			
			resp, err := client.Renew(ctx, req)
			cancel()

			if err != nil {
				log.Printf("Failed to renew lease: %v", err)
				continue
			}

			if !resp.Success {
				log.Printf("Lease renewal failed")
				// Try to re-register
				if err := g.register(); err != nil {
					log.Printf("Failed to re-register: %v", err)
				}
			}
		}
	}
}

// unregister removes the agent from the control plane
func (g *GrpcAgent) unregister() error {
	g.mu.Lock()
	if g.leaseID == "" {
		g.mu.Unlock()
		return nil
	}
	g.mu.Unlock()

	conn, err := grpc.NewClient(g.registryAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return fmt.Errorf("failed to connect to registry: %w", err)
	}
	defer conn.Close()

	client := registry.NewRegistryClient(conn)
	
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	req := &registry.AgentRegistration{Id: g.agent.GetID()}
	
	_, err = client.Unregister(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to unregister: %w", err)
	}

	log.Printf("Agent %s unregistered from control plane", g.agent.GetID())
	return nil
}

// Analyze implements the ConfidenceAgent.Analyze RPC
func (g *GrpcAgent) Analyze(ctx context.Context, req *confidence.AgentRequest) (*confidence.ConfidenceResult, error) {
	if req.TaskDescription == "" {
		return nil, status.Error(codes.InvalidArgument, "task description is required")
	}

	// Parse task data if provided
	var data interface{}
	if req.Data != nil {
		data = req.Data.AsMap()
	}

	// Call agent's analyze method
	result, err := g.agent.Analyze(ctx, req.TaskDescription, data)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "analysis failed: %v", err)
	}

	// Marshal result value
	valueJSON, err := json.Marshal(result.Value)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to marshal result: %v", err)
	}

	// Build response
	return &confidence.ConfidenceResult{
		ValueJson:     string(valueJSON),
		Confidence:    result.Confidence,
		AgentId:       g.agent.GetID(),
		Timestamp:     timestamppb.Now(),
		Reasoning:     result.Reasoning,
		Uncertainties: result.Uncertainties,
		Metadata:      result.Metadata,
	}, nil
}

// StreamAnalyze implements the ConfidenceAgent.StreamAnalyze RPC
func (g *GrpcAgent) StreamAnalyze(req *confidence.AgentRequest, stream grpc.ServerStreamingServer[confidence.ConfidenceResult]) error {
	// For now, just execute once and send result
	// TODO: Implement proper streaming
	
	resp, err := g.Analyze(stream.Context(), req)
	if err != nil {
		return err
	}

	return stream.Send(resp)
}

// GetCapabilities implements the ConfidenceAgent.GetCapabilities RPC
func (g *GrpcAgent) GetCapabilities(ctx context.Context, req *emptypb.Empty) (*confidence.Capabilities, error) {
	return &confidence.Capabilities{
		Capabilities: g.agent.GetCapabilities(),
	}, nil
}

// HealthCheck implements the ConfidenceAgent.HealthCheck RPC
func (g *GrpcAgent) HealthCheck(ctx context.Context, req *emptypb.Empty) (*confidence.Health, error) {
	health, err := g.agent.CheckHealth(ctx)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "health check failed: %v", err)
	}

	// Map status string to enum
	var pbStatus confidence.Health_Status
	switch health.Status {
	case "healthy":
		pbStatus = confidence.Health_HEALTHY
	case "degraded":
		pbStatus = confidence.Health_DEGRADED
	default:
		pbStatus = confidence.Health_UNHEALTHY
	}

	return &confidence.Health{
		Status:  pbStatus,
		Message: health.Message,
	}, nil
}

// ServeAgent is a convenience function to serve an agent
func ServeAgent(agent Agent, port int) error {
	grpcAgent := NewGrpcAgent(agent)
	
	if err := grpcAgent.Serve(port); err != nil {
		return err
	}

	// Wait for interrupt signal
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)
	
	<-sigCh
	log.Println("Shutting down agent...")
	
	return grpcAgent.Stop()
}

File: packages/sdk-rust/src/grpc_agent.rs
use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::Arc;
use std::time::Duration;

use async_trait::async_trait;
use tokio::signal;
use tokio::sync::Mutex;
use tokio::time::interval;
use tonic::{transport::Server, Request, Response, Status};
use tracing::{error, info, warn};

// Import generated proto types
use crate::generated::{
    confidence_agent_server::{ConfidenceAgent, ConfidenceAgentServer},
    registry_client::RegistryClient,
    Agent as ProtoAgent, ConfidenceResult, ExecuteRequest, ExecuteResponse,
    GetCapabilitiesResponse, HealthCheckResponse, RegisterRequest, RenewRequest,
    UnregisterRequest,
};

/// Trait that all Parallax agents must implement
#[async_trait]
pub trait Agent: Send + Sync + 'static {
    /// Get the agent's unique identifier
    fn get_id(&self) -> &str;

    /// Get the agent's human-readable name
    fn get_name(&self) -> &str;

    /// Get the agent's capabilities
    fn get_capabilities(&self) -> &[String];

    /// Get optional agent metadata
    fn get_metadata(&self) -> HashMap<String, String> {
        HashMap::new()
    }

    /// Perform the agent's main analysis task
    async fn analyze(
        &self,
        task: &str,
        data: Option<serde_json::Value>,
    ) -> Result<AnalyzeResult, Box<dyn std::error::Error>>;

    /// Check the agent's health
    async fn check_health(&self) -> Result<HealthStatus, Box<dyn std::error::Error>> {
        Ok(HealthStatus {
            status: "healthy".to_string(),
            message: Some("Agent is operational".to_string()),
        })
    }
}

/// Result of an agent's analysis
pub struct AnalyzeResult {
    pub value: serde_json::Value,
    pub confidence: f64,
    pub reasoning: Option<String>,
    pub uncertainties: Vec<String>,
    pub metadata: HashMap<String, String>,
}

/// Health status of an agent
pub struct HealthStatus {
    pub status: String, // "healthy", "degraded", "unhealthy"
    pub message: Option<String>,
}

/// gRPC-enabled agent wrapper
pub struct GrpcAgent<A: Agent> {
    agent: Arc<A>,
    port: u16,
    registry_addr: String,
    lease_id: Arc<Mutex<Option<String>>>,
    shutdown_tx: Arc<Mutex<Option<tokio::sync::oneshot::Sender<()>>>>,
}

impl<A: Agent> GrpcAgent<A> {
    /// Create a new gRPC-enabled agent
    pub fn new(agent: A) -> Self {
        let registry_addr = std::env::var("PARALLAX_REGISTRY")
            .unwrap_or_else(|_| "http://localhost:50051".to_string());

        Self {
            agent: Arc::new(agent),
            port: 0,
            registry_addr,
            lease_id: Arc::new(Mutex::new(None)),
            shutdown_tx: Arc::new(Mutex::new(None)),
        }
    }

    /// Serve the agent on the specified port
    pub async fn serve(mut self, port: u16) -> Result<(), Box<dyn std::error::Error>> {
        // Create gRPC service
        let service = GrpcAgentService {
            agent: self.agent.clone(),
        };

        // Build server
        let addr = format!("0.0.0.0:{}", port).parse::<SocketAddr>()?;
        let server = Server::builder()
            .add_service(ConfidenceAgentServer::new(service))
            .serve(addr);

        // Get actual address
        let actual_addr = server.local_addr();
        self.port = actual_addr.port();

        info!(
            "Agent {} ({}) listening on {}",
            self.agent.get_name(),
            self.agent.get_id(),
            actual_addr
        );

        // Register with control plane
        if let Err(e) = self.register().await {
            error!("Failed to register with control plane: {}", e);
            // Continue running even if registration fails
        }

        // Start lease renewal
        let agent_clone = self.agent.clone();
        let lease_id_clone = self.lease_id.clone();
        let registry_addr = self.registry_addr.clone();
        tokio::spawn(async move {
            renew_lease_loop(agent_clone, lease_id_clone, registry_addr).await;
        });

        // Setup shutdown handler
        let (shutdown_tx, shutdown_rx) = tokio::sync::oneshot::channel();
        *self.shutdown_tx.lock().await = Some(shutdown_tx);

        // Wait for shutdown signal
        let graceful = server.with_graceful_shutdown(async {
            let _ = shutdown_rx.await;
        });

        // Run server
        graceful.await?;

        // Unregister on shutdown
        if let Err(e) = self.unregister().await {
            error!("Failed to unregister: {}", e);
        }

        Ok(())
    }

    /// Register with the control plane
    async fn register(&self) -> Result<(), Box<dyn std::error::Error>> {
        let mut client = RegistryClient::connect(self.registry_addr.clone()).await?;

        let request = RegisterRequest {
            agent: Some(ProtoAgent {
                id: self.agent.get_id().to_string(),
                name: self.agent.get_name().to_string(),
                address: format!("localhost:{}", self.port),
                capabilities: self.agent.get_capabilities().to_vec(),
                metadata: self.agent.get_metadata(),
                status: 1, // HEALTHY
                last_seen: None,
            }),
        };

        let response = client.register(request).await?;
        let registration = response.into_inner().registration.unwrap();
        
        *self.lease_id.lock().await = Some(registration.lease_id);

        info!(
            "Agent {} registered with control plane",
            self.agent.get_id()
        );

        Ok(())
    }

    /// Unregister from the control plane
    async fn unregister(&self) -> Result<(), Box<dyn std::error::Error>> {
        let mut client = RegistryClient::connect(self.registry_addr.clone()).await?;

        let request = UnregisterRequest {
            agent_id: self.agent.get_id().to_string(),
        };

        client.unregister(request).await?;

        info!(
            "Agent {} unregistered from control plane",
            self.agent.get_id()
        );

        Ok(())
    }

    /// Stop the agent gracefully
    pub async fn stop(self) -> Result<(), Box<dyn std::error::Error>> {
        if let Some(tx) = self.shutdown_tx.lock().await.take() {
            let _ = tx.send(());
        }
        Ok(())
    }
}

/// Internal gRPC service implementation
struct GrpcAgentService<A: Agent> {
    agent: Arc<A>,
}

#[async_trait]
impl<A: Agent> ConfidenceAgent for GrpcAgentService<A> {
    async fn execute(
        &self,
        request: Request<ExecuteRequest>,
    ) -> Result<Response<ExecuteResponse>, Status> {
        let req = request.into_inner();
        let task = req.task.ok_or_else(|| Status::invalid_argument("task is required"))?;

        // Parse task data if provided
        let data = if task.data.is_empty() {
            None
        } else {
            Some(serde_json::from_str(&task.data).map_err(|e| {
                Status::invalid_argument(format!("invalid task data: {}", e))
            })?)
        };

        // Call agent's analyze method
        let result = self
            .agent
            .analyze(&task.description, data)
            .await
            .map_err(|e| Status::internal(format!("analysis failed: {}", e)))?;

        // Build response
        let response = ExecuteResponse {
            result: Some(ConfidenceResult {
                value_json: serde_json::to_string(&result.value).unwrap(),
                confidence: result.confidence,
                agent_id: self.agent.get_id().to_string(),
                timestamp: Some(prost_types::Timestamp::from(std::time::SystemTime::now())),
                reasoning: result.reasoning.unwrap_or_default(),
                uncertainties: result.uncertainties,
                metadata: result.metadata,
            }),
            error: None,
        };

        Ok(Response::new(response))
    }

    type StreamExecuteStream = tokio_stream::wrappers::ReceiverStream<Result<ExecuteResponse, Status>>;

    async fn stream_execute(
        &self,
        request: Request<ExecuteRequest>,
    ) -> Result<Response<Self::StreamExecuteStream>, Status> {
        // For now, just execute once and return
        // TODO: Implement proper streaming
        let (tx, rx) = tokio::sync::mpsc::channel(1);

        let response = self.execute(request).await?;
        let _ = tx.send(Ok(response.into_inner())).await;

        Ok(Response::new(tokio_stream::wrappers::ReceiverStream::new(rx)))
    }

    async fn get_capabilities(
        &self,
        _request: Request<()>,
    ) -> Result<Response<GetCapabilitiesResponse>, Status> {
        let response = GetCapabilitiesResponse {
            capabilities: self.agent.get_capabilities().to_vec(),
            expertise_level: 3, // EXPERT
            capability_scores: HashMap::new(),
        };

        Ok(Response::new(response))
    }

    async fn health_check(
        &self,
        _request: Request<()>,
    ) -> Result<Response<HealthCheckResponse>, Status> {
        let health = self
            .agent
            .check_health()
            .await
            .map_err(|e| Status::internal(format!("health check failed: {}", e)))?;

        let status = match health.status.as_str() {
            "healthy" => 1,   // HEALTHY
            "degraded" => 2,  // DEGRADED
            _ => 3,           // UNHEALTHY
        };

        let response = HealthCheckResponse {
            status,
            message: health.message.unwrap_or_default(),
        };

        Ok(Response::new(response))
    }
}

/// Lease renewal loop
async fn renew_lease_loop<A: Agent>(
    agent: Arc<A>,
    lease_id: Arc<Mutex<Option<String>>>,
    registry_addr: String,
) {
    let mut interval = interval(Duration::from_secs(30));

    loop {
        interval.tick().await;

        let lease = lease_id.lock().await.clone();
        if let Some(id) = lease {
            match renew_lease(&registry_addr, &id).await {
                Ok(renewed) => {
                    if !renewed {
                        warn!("Lease renewal failed, attempting to re-register");
                        // TODO: Re-register
                    }
                }
                Err(e) => {
                    error!("Failed to renew lease: {}", e);
                }
            }
        }
    }
}

/// Renew a lease with the registry
async fn renew_lease(
    registry_addr: &str,
    lease_id: &str,
) -> Result<bool, Box<dyn std::error::Error>> {
    let mut client = RegistryClient::connect(registry_addr.to_string()).await?;

    let request = RenewRequest {
        lease_id: lease_id.to_string(),
    };

    let response = client.renew(request).await?;
    Ok(response.into_inner().renewed)
}

/// Serve an agent with graceful shutdown
pub async fn serve_agent<A: Agent>(agent: A, port: u16) -> Result<(), Box<dyn std::error::Error>> {
    let grpc_agent = GrpcAgent::new(agent);

    // Setup signal handler
    let grpc_agent_clone = Arc::new(tokio::sync::Mutex::new(Some(grpc_agent)));
    let grpc_agent_signal = grpc_agent_clone.clone();

    tokio::spawn(async move {
        signal::ctrl_c().await.expect("Failed to listen for ctrl-c");
        info!("Shutting down agent...");
        
        if let Some(agent) = grpc_agent_signal.lock().await.take() {
            let _ = agent.stop().await;
        }
    });

    // Serve the agent
    if let Some(agent) = grpc_agent_clone.lock().await.take() {
        agent.serve(port).await?;
    }

    Ok(())
}